### ISSUE TITLE
CompositionLink fails a basic test

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#1475]
CompositionLink was introduced in #1473 but doesn't work for me...

Try this at home:
```
(use-modules (opencog))
(use-modules (opencog exec))

; define a function that returns x-1
(define minus-one
   (Lambda
      (Variable "$g")
      (Plus (Number -1) (Variable "$g"))))

(cog-execute!  (Put minus-one (Number 5)))
; prints (NumberNode "4.000000") as expected

; define a functin that returns x^2
(define square
   (Lambda
      (Variable "$x")
      (Times (Variable "$x")(Variable "$x"))))

(cog-execute!  (Put square (Number 5)))
; prints (NumberNode "25.000000") as expected

(define cmp (Composition minus-one square))

(cog-execute!  (Put cmp (Number 2)))
; Error: PutLink cannot handle CompositionLink

(cog-execute!  (Put cmp (List (Number 2))))
; This also fails.
```

**Comment by ngeiswei** [comment]
Thanks for the report. I'll look into this a bit later as I'm too eager to experiment with it. Basically for now it only returns the composed patterns when executing it `(cog-execute! (CompositionLink <g> <f>))`, which is enough for my current case.

**Comment by linas** [comment]
Well, but that does not work either: `(cog-execute! cmp )`  for the above throws an error too.

**Comment by ngeiswei** [comment]
Weird, **that** should work. I'll look into it tomorrow.

**Comment by linas** [comment]
I'm looking at some of this now.   I think it should be real easy.

**Comment by linas** [comment]
Its not easy.  For some reason, you are using ScopeLink with CompositionLink, and that is -- I'll say it again -- a fundamental mis-use of ScopeLink.   You can only compose LambdaLinks.   ScopeLinks fundamentally, as a matter of definition, cannot be composed!   We've been around this block multiple times, I just do not understand why ScopeLinks continue to get abused in this really horrible way.

**Comment by ngeiswei** [comment]
All right, then I'll derive *everything* from Lambda, ForAll, PutLink, GetLink, all of then, because composing them is ultimately very convenient.

**Comment by linas** [comment]
? But GetLink, BindLink, DualLink, PutLink cannot ever be derived from LambdaLink, because they are not composable!  That was kind of the point of inventing ScopeLink -- it tells what is not composable, while LambdaLink gives you exactly those things that *are* composable.     The ability to compose is exactly how Lambda and Scope differ from one-another.

**Comment by linas** [comment]
I mean -- there is no coherent way in which GetLink or BindLink could be composed.  That is the point.  Composing GetLink or BindLink would just result in garbage, some kind of nonsense expression that has no meaning.

**Comment by linas** [comment]
BTW, please note that PutLink also does function composition.  I'm still unclear on what CompostionLink is supposed to be doing, but I'm starting to wonder if maybe the reason you didn't use PutLink and invented something else was because PutLink was buggy.

I'm fixing the Times and plusLinks now, they have a bunch of bugs.

**Comment by ngeiswei** [comment]
PutLink bugginess isn't the reason I implemented ComposeLink. ComposeLink is a function constructor, it allows me to have for instance a URE rule take 2 functions as premises and built a new function as conclusion (with special properties dictated by the rule).

Regarding whether GetLink, etc are Lambda... Consider the fact that
```
GetLink <vardecl> <body>
```
is a shorthand for
```
GetLink
  Lambda <vardecl> <body>
```
GetLink is in fact a high level function that
1. take a function (Lambda <vardecl> <body>)
2. feed it into another function (in fact a MapLink or close) to it into a pattern
3. feed the whole think with an n-fold cartesian product of the atomspace (where n is the number of variables in `<vardecl>`)

`GetLink` implicitly represents all that. Because of that you can't compose 2 GetLink, unless you make all its arguments explicit (i.e the n-fold Cartesian product of the atomspace), then I suppose you could compose them, or rather you could say compose a MapLink with a GetLink, where the output of the GetLink does to feed the input of the MapLink (it's last argument, I mean).

You can unambiguously say however that
```
GetLink
  Compose
    Lambda X (Evaluation P X)
    Lambda X (Inheritance A X)
```
is equivalent to
```
GetLink
   Lambda
     X
     Evaluation P (Inheritance A X)
```

So we could utilize that to compose a Get with Lambda, for instance
```
Compose
  Get X (Evaluation P X)
  Lambda X (Inheritance A X)
```
To say the same thing as above.

I'm not necessary saying we should, but we certainly could.

Same thing for all other scope links, like ForAll, etc.

**Comment by linas** [comment]
> is a shorthand for
```
GetLink
  Lambda <vardecl> <body>
```
That's patently false. GetLink is not a LambdaLink. Its a ScopeLink.  It cannot be a Lambdalink because its not consistently beta-reducible. 

The problem with the analysis is that GetLink is not a "function". It is a pattern.  The problem with applying compistion to it is that the pattern to be searched for becomes ambiguous: its not clear how to treat the subgraph that you are trying to paste in there. What should the variable range over?  What pattern should actually be matched?

When you write 
```
GetLink
   Lambda
     X
     Evaluation P (Inheritance A X)
```
what does it mean to search for that?  Is X a free variable, here, or is it bound?  Am I supposed to search for 
```
GetLink
   Y
   Lambda
     X
     Evaluation P (Inheritance A X)
```
Am I supposed to search for 
```
GetLink
   X
   Lambda
     X
     Evaluation P (Inheritance A X)
```
? The use of a lambda in a pattern does not generate a well-founded expression.  It doesn't mean anything without a large pile of additional constraints and semantic interpretation.   you'd have to write raft of new code, and expand the concept of what it means to be a pattern in some very sophisticated ways before this could start making sense.

**Comment by linas** [comment]
Here is maybe a simpler example:  in your notiation, consider this form:
```
GetLink
  Compose
    Lambda X (Evaluation P X)
    Lambda X (Concept C)    ;; a function that is a constant
```
According to what you wrote, the beta-reduction of this is
```
(GetLink
    (Lambda (Concept C) (Evaluation P (Concept C)))
```
What does that mean? The `(Lambda (Concept C) ` doesn't make sense, and the Get is at most the Get of a constant pattern, trivially satisfiable.  You can replace `(Concept C)` by other atoms, the result is always nonsense.

**Comment by linas** [comment]
Anyway, I just merged a bunch of fixes to the arithmetic code.  The following is now possible:
```
(define minus-one
   (Lambda
      (Variable "$g")
      (Plus (Number -1) (Variable "$g"))))
(define square
   (Lambda
      (Variable "$x")
      (Times (Variable "$x")(Variable "$x"))))
(define pose (Put minus-one square))
(cog-execute! pose)
```
which returns
```
(PlusLink
   (LambdaLink
      (VariableNode "$x")
      (TimesLink
         (VariableNode "$x")
         (VariableNode "$x")
      )
   )
   (NumberNode "-1.000000")
)
```
which is almost correct, except that the Lambda needs to be commuted to the outermost level.  That is, it should have returned
```
   (LambdaLink
      (VariableNode "$x")
      (PlusLink
          (TimesLink
             (VariableNode "$x")
             (VariableNode "$x"))
          (NumberNode "-1.000000")))
```

When you say "function composition", the above seems to be what you want to happen; its not happening because Putlink is written too naively.

**Comment by ngeiswei** [comment]
Just to be clear, the composition of
```
Lambda X (Evaluation P X)
```
with
```
Lambda X (Concept C)
```
is
```
Lambda X (Evaluation P (Concept C))
```
You can test it. I don't know what I wrote that made you think it implies the contrary.

Regarding composability, I'll just say that anything that computes can be seen as a function, thus as something composable. The *only* constraint for composition is that the types of the composed functions are compatible, in particular, it means that in `Compose <g> <f>` the co-domain of f has to inherit from the domain of g. So if we formulate `GetLink` in a completely functional manner (perhaps using state monads, etc), then we should be able to construct well-founded compositions. I'm not suggesting to do it, cause it might not be worth the pain, not in the immediately future anyway, I'm just saying that it's possible.

**Comment by ngeiswei** [comment]
Yes, and you have shown, `PutLink` is almost `ComposeLink`. I suppose we could implement `ComposeLink` with `PutLink` plus other operators, I'm open to it, but for practical immediate purposes `ComposeLink` it doing what I need.

**Comment by linas** [comment]
OK, I think I see the general idea. I have to think about it a bit.  I think I located one of the conceptual difficulties.

In "ordinary" lambda calculus, it is common to have expressions like  
```
lambda x. (foo x bar lambda y. (baz y bork))
```
This is a valid lambda expression, but it is not a valid scope expression -- scopes have to have all variable definitions out in front.   Now, the above can be converted into a valid scope format like so:
```
lambda xy. (foo x bar  baz y bork)
```
that is, by "pulling" all lambdas in all subexpressions all the way over to the left.  There does not seem to be a name for the operation of "pulling all lambdas all the way to the left".  (I spent the last hour googling this)  (I propose calling this "shift conversion" or "scope conversion").

Anyway, the second form **is** a valid "pattern", and can be used for searching. The first form is not a valid pattern, and cannot be used for searching.  The intent of the ComposeLink seems to be to perform the beta reduction and also the "scope conversion", so that the result is a valid pattern.

**Comment by linas** [comment]
I guess that "pulling all lambdas all the way to the left" should be called "prenex conversion".

**Comment by linas** [comment]
https://en.wikipedia.org/wiki/Prenex_normal_form -- I'm thinking of maybe creating a `PrenexLink` which is a ScopeLink that has to always forced to be in prenex form.   Thus, all search patterns are always in prenex form...

**Comment by linas** [comment]
Nil, what are you using ComposeLink for?  One issue is that `types.script` says that the ComposeLink is a kind-of FunctionLink, and composes FunctionLinks.  Now, FunctionLinks are a kind-of FreeLink ...  and FreeLinks don't have any variable declarations.   And thats fine: a way of composing FunctionLinks.

But then, all of your examples involve Lambda's and lambdas are not FunctionLinks, they are not FreeLinks, they are ScopeLinks, and always have variable declarations.  And the way in which one composes ScopeLinks has to be different than the way one composes FunctionLinks; they cannot work the same way.  For example, to compose a ScopeLink, you'd need to perform type checking.  

By contrast, a  FunctionLink has two kinds of variables: the implicit ones, and the free ones.  To keep things simple, think of PlusLink as a "typical" FunctionLink.

So: PlusLink has an infinite number of implicit variables, and its always implicitly and automatically composable.  It has zero free variables: there's no way to say "substitute for x in PlusLink", since there is no x.   The only way to have a free variable is to write:
```
Plus
    Number 2
    Variable x
```
Note that x is free, and is not typed, nor is it declared.   The only way to declare it would be to write
```
Lambda
     VariableType  x Number
     Plus
         Number 2
         Variable x
```
However, a Lambda is not a Function!  Its a Scope!  

I suspect that its a mistake to say that "ComposeLink composes functions".. I think you maybe meant to say its a kind of ScopeLink, not a kind of FunctionLink.   I can't quite figure out what its supposed to actually be, or what its supposed to do.  I mean I do at some abstract level, but its just bizarro at the C++ code level.

**Comment by linas** [comment]
For example you write:  
```
// Take an n-ary function g as first outgoing,
// a ListLink of n m-ary functions f1 to fn, as second outgoing, and
// produce the m-ary function g(f1(x1,...,xm),...,fn(x1,...,xm)). 
```
The problem with this is that FunctionLinks don't have an arity, so it doesn't make sense to say "take an m-ary function" because m doesn't exist, cannot be defined for a FunctionLink.   Only ScopeLinks have an arity.

**Comment by linas** [comment]
FunctionLinks are not composable.   For example PlusLink and TimesLink are both FunctionLinks.  But you can't say:
```
ComposeLink
     PlusLink
     TimesLink
```
because.. PlusLink has no arity, TimesLink has no arity, .. this is just a nonsense expression.  What you can do is to compose ScopeLinks (provided you do the type checking) and you can compose specific instances of FreeLinks.  You just cannot compose FunctionLinks in general.

**Comment by linas** [comment]
The new ScopeLink with the alpha conversion and partial substitution also doesn't make sense.  It violates the idea of modular, object-oriented coding: "do one thing, and do one thing well".  I'm trying to invent a new kind of link that will do the things you are trying to do with ScopeLink, but refactors it into it's own atom type.  

The issue here is that the `partial_substitute()` method is not a virtual method, and so it can't be overloaded to handle the different kinds of substitutions that different kinds of atom types need.  I'd rather have each kind of atom figure out what it means to "substitute" or "compose", and handle that particular style of composition that it actually has to do.

**Comment by linas** [comment]
The other problem with `partial_substitute` is that  (almost?) everything that `partial_substitute()` does is already done by the Instantiator class and/or the Substitutor class-- except its done slightly differently.  So it does not quite behave the same way. 

Did you forget that the Instantiator/Substitutor class existed, when you created `partial_substitute`?  Did you look at Instantiator, and decide it was too hard to alter, to make it doe what you want to do?  I mean, we now do beta-reduction in maybe 6 or 7 different places, and they are all similar-but-slightly different, and I'm having a hard time figuring out what the differences are.  These are:

* Instantiator class
* Variables class
* PutLink class
* ComposeLink class
* ScopeLink class
* Substitutor class

Its fucked up that we have six ways of doing the same thing.   I'm going to start to try to somehow refactor all of these, because I can't deal with debugging, its just getting too hard.

**Comment by linas** [comment]
I'm thinking of create a SubstituteLink that derives from ScopeLink, and will move all the substitution methods from ScopeLink to the SubstituteLink.   After this, I plant to have two types of substitution: ordinary lambda substitution and prenex substitution.  I believe that  what you are trying to do with ComposeLink is to perform beta-reduction, but then always put the result into prenex form. 

The PutLink can then do what it does today, but the result will differ depending on whether the links is a LambdaLink or a PrenexLink.   The GetLink, BindLink, DualLink, SatisfiabilitlityLinks will all be PrenexLinks. 

I'm still unclear on where that leave ComposeLink, because I don't understand the intended use case.

**Comment by ngeiswei** [comment]
On 12/16/2017 02:41 AM, Linas Vepštas wrote:
> Nil, what are you using ComposeLink for? One issue is that 

To build bigger patterns out of smaller patterns, as needed by the WIP 
URE-based pattern miner. It allows to express specialization (in fact 
composition) at the top of the hypergraph rather buried arbitrarily deep 
in the pattern.

Maybe this rule will make it clearer why I need this (if you can 
allocate the 10 minutes it may take to understand it) 
https://github.com/ngeiswei/opencog/blob/experimental-pattern-miner/opencog/learning/experimental-pattern-miner/ure-based/backwardchainer/specialization-rule.scm

> |types.script| says that the ComposeLink is a kind-of FunctionLink, and 
> composes FunctionLinks. Now, FunctionLinks are a kind-of FreeLink ... 
> and FreeLinks don't have any variable declarations. And thats fine: a 
> way of composing FunctionLinks.

ComposeLink is a FreeLink indeed, it has no scope, however it's 
arguments are (usually?) scopes or lambdas.

I made it inherit from FunctionLink because it is in spirit a function, 
albeit a higher order one. Maybe it should just inherit Link.

Note however that the result of a ComposeLink, after execution (or 
perhaps it should be called reduction, if there's a difference) is a 
scope or lambda, as shown in its unit tests 
https://github.com/opencog/atomspace/blob/master/tests/atoms/ComposeLinkUTest.cxxtest

> By contrast, a FunctionLink has two kinds of variables: the implicit 
> ones, and the free ones.

Implicit ones? Sorry, I don't understand.

> To keep things simple, think of PlusLink as a 
> "typical" FunctionLink.
> 
> So: PlusLink has an infinite number of implicit variables, and its 
> always implicitly and automatically composable. It has zero free 
> variables: there's no way to say "substitute for x in PlusLink", since 
> there is no x. The only way to have a free variable is to write:
> 
> |Plus Number 2 Variable x | >
> Note that x is free, and is not typed, nor is it declared. The only way 
> to declare it would be to write
> 
> |Lambda VariableType x Number Plus Number 2 Variable x |
> 
> However, a Lambda is not a Function! Its a Scope!

I think I see the problem. A PlusLink, while "unused" is a actually a 
lambda. But as soon as it has outgoing(s) it has essentially been 
applied. In Lambda calculus terms, a uninstantiated PlusLink is a an 
abstraction (so it has a lambda), but as soon as it is put in an 
AtomSpace it amounts to putting a application over it, and its outgoings 
are the arguments of that application.

BTW, this issue is somewhat related to 
https://github.com/opencog/atomspace/issues/1456.

OK, for instance, let say we have a GroundedSchema called "Plus" such that

Equal
   ExecutionOutputLink
     GroundedSchema "Plus"
     List 2 3
   Plus 2 3

Then GroundedSchema "Plus" would essentially be a lambda. However I 
think GroundedSchema "Plus" is different than a nullary PlusLink because 
the latter is has been applied with an empty argument list, but still 
applied.

See what I mean?

So basically what we need is a way to be able to talk about PlusLink a 
function object without using it, maybe the GroundedSchema "Plus" could 
be the hardwired associated function object.

**Comment by ngeiswei** [comment]
On 12/16/2017 02:46 AM, Linas Vepštas wrote:
> |// Take an n-ary function g as first outgoing, // a ListLink of n m-ary 
> functions f1 to fn, as second outgoing, and // produce the m-ary 
> function g(f1(x1,...,xm),...,fn(x1,...,xm)). |
> 
> The problem with this is that FunctionLinks don't have an arity, so it 
> doesn't make sense to say "take an m-ary function" because m doesn't 
> exist, cannot be defined for a FunctionLink. Only ScopeLinks have an arity.

Yeah FunctionLinks like PlusLink can take an arbitrary number of 
arguments, but the functions I consider so far are lambdas and have a 
determined number of arguments.

**Comment by ngeiswei** [comment]
On 12/16/2017 03:01 AM, Linas Vepštas wrote:
> FunctionLinks are not composable. For example PlusLink and TimesLink are 
> both FunctionLinks. But you can't say:
> 
> |ComposeLink PlusLink TimesLink |
> 
> because.. PlusLink has no arity, TimesLink has no arity, .. this is just 
> a nonsense expression. What you can do is to compose ScopeLinks 
> (provided you do the type checking) and you can compose specific 
> instances of FreeLinks. You just cannot compose FunctionLinks in general.

It doesn't say that ComposeLink can compose function links (BTW, maybe 
FunctionLink should be renamed ApplicationLink, or at least it should 
make it clear that the existence of such a link in the atomspace 
represents a function application), it says that ComposeLink *itself* is 
a function link.

**Comment by ngeiswei** [comment]
On 12/16/2017 04:02 AM, Linas Vepštas wrote:
> The issue here is that the |partial_substitute()| method is not a 
> virtual method, and so it can't be overloaded to handle the different 
> kinds of substitutions that different kinds of atom types need. I'd 
> rather have each kind of atom figure out what it means to "substitute" 
> or "compose", and handle that particular style of composition that it 
> actually has to do.

Hmm, it doesn't hurt to make it virtual, but it's unclear to me that we 
want particular composition, but maybe we do.

**Comment by ngeiswei** [comment]
On 12/16/2017 04:18 AM, Linas Vepštas wrote:
> Did you forget that the Instantiator/Substitutor class existed, when you 
> created |partial_substitute|? Did you look at Instantiator, and decide 
> it was too hard to alter, to make it doe what you want to do?

I was afraid to break the pattern matcher in major ways, since they all 
re-used the main substitution function `Variables::substitute_nocheck`, 
I felt it's OK in terms DRY principle.

Also the ComposeLink (and formerly ScopeLink) substitutions are more 
sophisticated with quotations handling (I really need to better document 
that, let me do that next), but maybe that can be ported to the other 
substition methods.

**Comment by ngeiswei** [comment]
On 12/16/2017 06:41 AM, Linas Vepštas wrote:
> I'm thinking of create a SubstituteLink that derives from ScopeLink, and 
> will move all the substitution methods from ScopeLink to the 
> SubstituteLink. After this, I plant to have two types of substitution: 
> ordinary lambda substitution and prenex substitution. I believe that 
> what you are trying to do with ComposeLink is to perform beta-reduction, 
> but then always put the result into prenex form.
> 
> The PutLink can then do what it does today, but the result will differ 
> depending on whether the links is a LambdaLink or a PrenexLink. The 
> GetLink, BindLink, DualLink, SatisfiabilitlityLinks will all be PrenexLinks.
> 
> I'm still unclear on where that leave ComposeLink, because I don't 
> understand the intended use case.

But the user should be able to turn its substituted lambda to a prenex 
form, whether or not it inherits from PrenexLink, right?

For instance I do intent to use ComposeLink over ImplicationScopeLink, 
specifically

```
Compose
   ImplicationScope
     ...
   Lambda
     ...
```
in order to define specialized versions of that ImplicationScope.

I simply that whether you want or not a prenex form is gonna depend on 
the use, I think.

**Comment by linas** [comment]
OK, well, for some reason, I thought it was a good idea to re-organize the code to more clearly separate out distinct concepts into their own atom-types.  Its more difficult than I thought.  I'm not done yet...  more later, am trying to keep the above in mind.

**Comment by linas** [comment]
So far, the points are these:

* What FunctionLinks implement is often called "delta-reduction" -  the FunctionLink is replaced by some value.  In general comp-sci (viz other programming languages), this is called "execution" or "evaluation" or "apply" but I'm trying to be more precise, because the failure to be precise is causing design issues.

* When talking about traditional lambda calc, there are three different kinds of "substitution": alpha conversion (substituting one variable for another) beta-reduction (substituting a constant value for a variable and dropping the lambda) and eta conversion (dropping a lambda entirely, when the variable it binds is appended)  The old code did beta only. Your new code does beta and sometimes alpha (and quotes) but never eta.

* The "difference" between a FunctionLink and a LambdaLink is eta-conversion. In traditional lambda calc, it is this. The expression  Then `lambda x. cosine x` can be eta-reduced to just `cosine` and going in the opposite direction, `cosine` can be eta-abstracted into `lambda x. cosine x`.  Both are kind-of "the same"-ish, thank to eta-conversion. So what the heck is `cosine`? Its an atom that, when given a numeric value, can be delta-reduced to a different numeric value. 

*In many of your emails, you say things like "blah blah is a lambda" and I scream bloody murder because "blah blah is not a lambda".  I realize now that you meant to say "blah blah is eta-convertable to a lambda". In fact, many traditional programming languages thrive on making eta-conversion implicit or invisible or tacit.  It allows one to very conveniently "pretend" that lambda expressions are "functions". They're not, they are just eta-convertible to one-another.

* This makes clear what "apply" is: `apply f x` means "eta abstract f then beta-reduce".  If `x` is a numeric value, and `f` is a numeric function, then  `apply f x` means "eta abstract f then delta-reduce".  If `x` is a variable, then `apply f x` means "just eta-abstract f with x and nothing more".

* I'm trying to make LambdaLink behave as lambda-calculus-like as possible. That means many things: no implicit/automatic eta-conversion, no implicit/automatic beta-conversion. its not automatically in prenex form.

* The wikipedia article on lambda calculus is chock-full of errors and confusions and is not reliable.

* However, it is "well-known" that alpha-conversion sucks, and there have been numerous techniques invented for avoiding it, for example, de-bruin indexes, combinators ... Thus I am pained when I see explicit alpha-conversion being implemented in the c++ code.  I keep thinking that it can be avoided if only one was more clever.

* The PatternLink does always have to be in prenex form.

Anyway, that's a summary of how I'm currently thinking about the problems.  

There is a meta-issue here: much of what is written in the c++ code are in fact ad-hoc, poorly-spec'ed "rules of inference" and ideally, we could/should have them be rules in the URE instead. However, its clear that we can't do this yet, because whenever we try, quote and unquote are needed.   My personal impression is that quote and unquote are evil, and are needed only because somehow we mis-designed atomese.  I have not yet figured out what the "right" solution is.

**Comment by ngeiswei** [comment]
Thanks, Linas, it does make it clearer (I had completely forgotten about eta-conversion).

Regarding Quote/Unquote, one of the problems I see is that it has 2 different uses
1. Turn variable into "text" for pattern matching ("text" is within the terminology of the pattern mining literature to indicate data, as opposed to patterns).
2. Prevent execution (I didn't find any formal definitions of delta-reduction so I'm hesitant to use it, I welcome some pointers if you have).

We also have `DontExecLink` for the second usage with the difference that it isn't consumed by the pattern matcher / execution.

I don't know what is the right way to approach that either... I guess we could attempt to experiment with links like `BindQuoteLink` instead of `(Quote (Bind (Unquote ...`, etc, we might learn something.

**Comment by linas** [comment]
I would rather not invent "BindQuoteLink" or stuff like that; I would rather have a syntax where its always clear what part is "text" and what part is "pattern".   I don't understand how we got into this situation of needing to have quotes. I keep thinking that we did something wrong, somewhere along the way. To fix it, I would need a strong example of how we use quotes today, that shows why they are really needed.

**Comment by linas** [comment]
googling this is not easy. The most authoritative reference is the last one, from a book.
http://foldoc.org/delta+reduction
http://barrywatson.se/lsi/lsi_delta_reduction.html
http://kronopath.net/nyu/pl-notes/Mar-9-Lecture/
https://www.cise.ufl.edu/class/cop5555su11/rpal/lambda.calculus.pdf
https://books.google.com/books?id=Ojp6YuDI_AsC&pg=PA87&lpg=PA87&dq=delta+rule+lambda&source=bl&ots=XtIWwR80ai&sig=_xZO7rBcf5GMsdk69dZF2IxjcvM&hl=en&sa=X&ved=0ahUKEwiemO6w0pXYAhVFSSYKHWYMCcw4ChDoAQhSMAc#v=onepage&q=delta%20rule%20lambda&f=false

**Comment by linas** [comment]
Closing, this won't go anywhere.
